src/Tools/CodeGenerator/Featurizers.yaml

```diff
@@ -639,6 +639,75 @@ featurizers:
       - input_type: TInputDouble
         output_type: double
 
+    # ----------------------------------------------------------------------
+  - name: MaxHorizonFeaturizer
+    estimator_name: MaxHorizonEstimator
+    release_version: 0.4.0
+    num_output_columns: 1
+    # num_output_columns=1 is required by schema but not makes sense in this featurizer
+    # consider a potantial change in schema if another featurizer with similar concepts occurs
+
+    description: |-
+
+      This Featurizer will be completely implemented in Frameworks(ORT and ML.NET).
+
+      This Transformer adds rows according to maxHorizon and also appends a column
+      which is horizon_origin. The overall number of output rows is maxHorizon times
+      the input's number of rows.
+
+      The type of input and output is abstracted to Tabular which corresponding to
+      Tensor in OnnxRuntime and IDataView in ML.NET.
+
+      C++-style pseudo signature:
+        Tabular execute(Tabular const &value)
+
+      Examples:
+
+        The Transformer duplicate rows according to maxHorizon, and adding one
+        column called horizon_origin. The following example sets maxHorizon as 2.
+
+        Input:
+          +------------+-------+-------+
+          | date       | grain | sales |
+          +============+=======+=======+
+          | 2017-01-01 | A     |  0    |
+          +------------+-------+-------+
+          | 2017-02-01 | A     |  1    |
+          +------------+-------+-------+
+          | 2017-03-01 | A     |  2    |
+          +------------+-------+-------+
+          | 2017-04-01 | A     |  3    |
+          +------------+-------+-------+
+
+        Output:
+          +------------+-------+-------+----------------+
+          | date       | grain | sales | horizon_origin |
+          +============+=======+=======+================+
+          | 2017-01-01 | A     | 0     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-01-01 | A     | 0     | 2              |
+          +------------+-------+-------+----------------+
+          | 2017-02-01 | A     | 1     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-02-01 | A     | 1     | 2              |
+          +------------+-------+-------+----------------+
+          | 2017-03-01 | A     | 2     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-03-01 | A     | 2     | 2              |
+          +------------+-------+-------+----------------+
+          | 2017-04-01 | A     | 3     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-04-01 | A     | 3     | 2              |
+          +------------+-------+-------+----------------+
+
+    configuration_params:
+      - type: uint32

```

Comment:
By changing this to uint32, it implies that we can have a horizon of 2^32. Is this really what we want? My preference is:

1) uint16
2) uint8
3) uint32

Reply:
Can you explain why is this your preference? It really can not be a uint8 as we need to use it in arithmetic operations such as a counter. We'd need to cast it to one of the types.  Why not use the natural datatype for which the processor has built in instructions?

It will also cause padding in the class anyway.

Reply:
This is the type that will be expressed in the estimator's constructor as the number of rolling windows to calculate for each "row". When we restrict the type to something smaller, we are communicating to the caller the range of acceptable values. I look at it as a form of documentation to the caller.

In the internal Estimator implementation, we can convert this to whatever type is appropriate based on how it is used (for example, the class might store this value as uint32 even though the incoming type is something smaller).

Reply:
We have a bit of docs about the max horizon parameter: https://docs.microsoft.com/en-us/azure/machine-learning/how-to-configure-auto-train

> `max_horizon`: Defines the length of time you want to predict out based on the periodicity of the training data. For example if you have training data with daily time grains, you define how far out in days you want the model to train for.

 While we expect most users would be ok with UINT8, a subset of users have larger needs.

Let's assume your interval of data is one day. A UINT8 would let you predict 255 days in the future. It seems reasonable for someone to ask, "how many oranges will I sell for the next year", which a UINT8 would not allow.

A reasonable upper limit for this is also beyond a UINT16 for similar reasons. Let's assume your data arrives every second. A UINT16 would let you predict 65,535 seconds in the future (18 hours). It also seems reasonable for someone to ask, "what will my traffic be for the next day", which a UINT16 would not allow.

A UINT32 with data arriving every second, would be 136 years. Certainly more than we expect users to request.

Hence my recommendation for a UINT32 to store the max horizon.

Reply:
Thanks, Justin - I'm convinced!

Reply:
Correct, this is a multiplier of the period of the user's data, and it limits how far into the future the user can predict to `<= period * max_horizon`.

While most users would do ok with only a UINT8, some may need larger than a UINT16; hence the UINT32 is best.

Reply:
That's fine, we can make it uint32. 

However, can you go into a bit more detail about scenarios when an uint32 would be necessary? I want to make sure that we have a good understanding of those as we are talking about a large enough data set that it might alter other designs. Thanks, Justin.